### PR TITLE
Software module metadata available to targets (in DMF and DDI)

### PR DESCRIPTION
**Pull request by kaizimmerm**
Sometimes devices need additional metadata of the module (i.e. next to static information like type.key and version etc.).

With this change metadata elements can be marked as "targetVisible" and as result become part of Update DMF message or deploymentBase DDI resource.

Main changes:
* SoftwareModuleMetadata in repo has now a real builder.
* Mgmt UI/API, DDI and DMF extended
* Metadata UI code refactored to reduce duplicate code between DS and SM Metadata management.

### CODE REVIEW

File: hawkbit-ddi-resource/src/test/resources/ddi-test.properties [hunk 1]
```diff
@@ -20,3 +20,5 @@ spring.http.multipart.max-file-size=5MB
 hawkbit.server.security.dos.maxStatusEntriesPerAction=100
 hawkbit.server.security.dos.maxAttributeEntriesPerTarget=10
 # Quota - END
+
+logging.level.org.eclipse.hawkbit.rest.util.MockMvcResultPrinter=DEBUG
```
Comment (kaizimmerm) [inline hawkbit-ddi-resource/src/test/resources/ddi-test.properties]: remove

File: hawkbit-dmf/hawkbit-dmf-api/src/main/java/org/eclipse/hawkbit/dmf/json/model/DmfSoftwareModule.java [hunk 2]
```diff
@@ -36,6 +33,9 @@
     private String moduleVersion;
     @JsonProperty
     private List<DmfArtifact> artifacts;
+    @JsonProperty
+    @JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (kaizimmerm) [inline hawkbit-dmf/hawkbit-dmf-api/src/main/java/org/eclipse/hawkbit/dmf/json/model/DmfSoftwareModule.java]: Remove

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/ControllerManagement.java [hunk 3]
```diff
@@ -72,7 +73,10 @@
      * @return {@link SoftwareModule} identified by ID
      */
     @PreAuthorize(SpringEvalExpressions.IS_CONTROLLER)
-    Optional<SoftwareModule> getSoftwareModule(@NotNull final Long moduleId);
+    Optional<SoftwareModule> getSoftwareModule(@NotNull Long moduleId);
+
+    @PreAuthorize(SpringEvalExpressions.IS_CONTROLLER)
+    List<SoftwareModuleMetadata> findTargetVisbileMetaDataBySoftwareModuleId(@NotNull Long moduleId);
```
Comment (kaizimmerm) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/ControllerManagement.java]: Javadocs

Comment (kaizimmerm) [reply hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/ControllerManagement.java]: Typo Visible

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/model/MetaData.java [hunk 4]
```diff
@@ -15,6 +15,15 @@
  *
  */
 public interface MetaData extends Serializable {
+    /**
+     * Maximum length of metadata key.
+     */
+    int KEY_MAX_SIZE = 256;
```
Comment (kaizimmerm) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/model/MetaData.java]: Check value.

File: hawkbit-core/src/main/java/org/eclipse/hawkbit/repository/SoftwareModuleMetadataFields.java [hunk 5]
```diff
@@ -24,7 +24,12 @@
     /**
      * The key field.
      */
-    KEY("key");
+    KEY("key"),
+
+    /**
+     * The target visible field.
+     */
+    TARGETVISIBLE("targetVisible");
```
Comment (kaizimmerm) [inline hawkbit-core/src/main/java/org/eclipse/hawkbit/repository/SoftwareModuleMetadataFields.java]: Add to docs

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/detailslayout/SoftwareModuleMetadataDetailslayout.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+import org.springframework.data.domain.PageRequest;
+import org.springframework.util.CollectionUtils;
+
+import com.vaadin.ui.UI;
+
+/**
+ * SoftwareModule Metadata details layout.
+ *
+ */
+public class SoftwareModuleMetadataDetailslayout extends AbstractMetadataDetailslayout {
```
Comment (blomark) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/detailslayout/SoftwareModuleMetadataDetailslayout.java]: Typo: layout -> Layout

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/detailslayout/AbstractMetadataDetailslayout.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import com.vaadin.data.util.IndexedContainer;
+import com.vaadin.ui.Button;
+import com.vaadin.ui.Table;
+import com.vaadin.ui.themes.ValoTheme;
+
+/**
+ * Abstract metadata tab for entities.
+ *
+ */
+public abstract class AbstractMetadataDetailslayout extends Table {
```
Comment (blomark) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/detailslayout/AbstractMetadataDetailslayout.java]: Typo: layout -> Layout

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/builder/SoftwareModuleMetadataCreate.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+     * @return updated builder instance
+     */
+    SoftwareModuleMetadataCreate value(@Size(min = 1, max = MetaData.VALUE_MAX_SIZE) @NotNull String value);
+
+    /**
+     * @param visible
+     *            for {@link SoftwareModuleMetadata#isTargetVisible()}
+     * @return updated builder instance
+     */
+    SoftwareModuleMetadataCreate targetVisible(Boolean visible);
```
Comment (michahirsch) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/builder/SoftwareModuleMetadataCreate.java]: `Boolean` wrapper type? I guess we could use the primitive type in the builder API, right?

File: hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/builder/SoftwareModuleMetadataUpdate.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+     * @return updated builder instance
+     */
+    SoftwareModuleMetadataUpdate value(@Size(min = 1, max = MetaData.VALUE_MAX_SIZE) @NotNull String value);
+
+    /**
+     * @param visible
+     *            for {@link SoftwareModuleMetadata#isTargetVisible()}
+     * @return updated builder instance
+     */
+    SoftwareModuleMetadataUpdate targetVisible(Boolean visible);
```
Comment (michahirsch) [inline hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/builder/SoftwareModuleMetadataUpdate.java]: as above => `Boolean` wrapper type? I guess we could use the primitive type in the builder API, right?

**Pull request merged**